### DISCUSSION TITLE
79. 인덱스에서 선택도(Selectivity)가 무엇인가요?

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#81] [General]
인덱스에서 선택도(Selectivity)가 무엇인가요?

**Comment by linirini** [comment]
인덱스 선택도는 데이터베이스 인덱스의 효율성을 나타내는 척도로, 쿼리 조건에 따라 얼마나 많은 행이 반환될 것인지를 나타낸다. 간단히 말해, 조건을 만족하는 행의 비율을 의미한다.
데이터베이스의 쿼리 옵티마이저는 선택도를 참고하여 인덱스를 사용할지 또는 전체 테이블 스캔을 할지 결정한다.

0 <= (선택도 = 조건에 의해 반환되는 행의 수 / 전체 행의 수) <= 1
- 0에 가까울수록: 반환되는 행의 수가 적어 인덱스 효율이 높음.
- 1에 가까울수록: 반환되는 행의 수가 많아 인덱스 효율이 낮음.

**인덱스 효율성**
- 선택도가 낮은 경우(즉, 조건이 데이터의 작은 부분만 반환하는 경우), 인덱스를 사용하는 것이 효율적
- 선택도가 높은 경우, 인덱스를 사용해도 대부분의 데이터를 읽어야 하므로 테이블 스캔이 더 나을 수 있다.

**Reply by cutehumanS2** [reply]
수식에서 분자와 분모가 반대로 되어 수정해 두었습니다.

**Comment by cutehumanS2** [comment]
인덱스에서 선택도에 대한 다른 관점도 있어 의견 추가로 남깁니다. 😸
<br>
    
인덱스에서 **선택도**는 카디널리티의 개념과 연결해 볼 수 있습니다.
**카디널리티**란 인덱스 키 값 중 **유니크한 값의 수**를 의미하며, 카디널리티가 높을수록 선택도가 좋아진다고 해석할 수 있습니다. 
보통 선택도가 높을수록(즉, 유니크한 값이 많을수록) 검색 대상이 줄어들어 인덱스를 사용하는 것이 더 효율적입니다. 

_+) 선택도가 낮더라도 정렬이나 그룹핑 작업에서는 인덱스가 더 효율적인 경우가 있습니다._

<br>

[참고한 자료]
- https://www.programmerinterview.com/database-sql/selectivity-in-sql-databases/
- 책 <Real MySQL 8.0 (1권)> - 8.3 B-Tree 인덱스
